lightning/src/ln/onion_utils.rs

```diff
@@ -229,12 +238,45 @@ pub(super) fn construct_onion_packet_bogus_hopdata<HD: Writeable>(payloads: Vec<
 	let mut chacha = ChaCha20::new(&prng_seed, &[0; 8]);
 	chacha.process(&[0; ONION_DATA_LEN], &mut packet_data);
 
-	construct_onion_packet_with_init_noise(payloads, onion_keys, packet_data, associated_data)
+	construct_onion_packet_with_init_noise::<_, _>(
+		payloads, onion_keys, FixedSizeOnionPacket(packet_data), Some(associated_data))
+}
+
+/// Since onion message packets and onion payment packets have different lengths but are otherwise
+/// identical, we use this trait to allow `construct_onion_packet_with_init_noise` to return either
+/// type.
+pub(crate) trait Packet {
+	type D;
+	fn new(pubkey: PublicKey, hop_data: Self::D, hmac: [u8; 32]) -> Self;
+}
+
+// Needed for rustc versions older than 1.47 to avoid E0277.
+pub(crate) struct FixedSizeOnionPacket([u8; ONION_DATA_LEN]);
+
+impl AsMut<[u8]> for FixedSizeOnionPacket {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0
+	}
+}
+
+impl Packet for msgs::OnionPacket {
```

Comment:
I may have been pushing for having `onion_message/packet.rs` depend on `ln/onion_utils.rs` rather than having both depend on each other, FWIW. Not against having limited mutual dependencies where it makes sense.

Reply:
Or more generally between `onion_message` module (rather than specifically `packet` submodule) and `onion_utils`.